src/mapgen/mg_biome.cpp

```diff
@@ -297,7 +297,7 @@ Biome *BiomeGenOriginal::calcBiomeFromNoise(float heat, float humidity, v3s16 po
 	// Carefully tune pseudorandom seed variation to avoid single node dither
 	// and create larger scale blending patterns similar to horizontal biome
 	// blend.
-	const u64 seed = pos.Y + (heat + humidity) * 0.9f;
+	const u64 seed = std::abs(pos.Y + (heat + humidity) * 0.9f);
```

Comment:
I don't think throwing `abs` on this is the right thing to do, since it will be inconsistent with previous behavior.

This change in its current form has the potential to break mapgen backwards compatibility (as in: old mapblocks may be inconsistent with newly generated mapblocks).

We would need to make sure that we preserve the old behavior, which is tricky however (see e.g. the generated assembly [on GodBolt](https://godbolt.org/z/MKnYenaT6)).

Reply:
It seems that the unintended behavior has become a feature. In this case, since they are not really undefined behavior, I would simply remove these patches.

Reply:
> It seems that the unintended behavior has become a feature.

Indeed.

> In this case, since they are not really undefined behavior, I would simply remove these patches.

Feel free to add a comment along the lines of `// FIXME: This is technically undefined behavior, but it is not trivial to provide an equivalent implementation without UB, which is required for backwards compatibility`.

Reply:
Is not assigning signed negative number into unsigned number also a defined behavior?

Reply:
Yes, converting a signed integer to unsigned should be well-defined.

But this is converting a possibly negative float to an unsigned integer, which is undefined behavior [according to this source](https://embeddeduse.com/2013/08/25/casting-a-negative-float-to-an-unsigned-int/).

Reply:
We could also try the proposed fix of first casting to `s64`, assuming `pos.Y + (heat + humidity) * 0.9f` does not violate `s64` integer bounds (which I think it shouldn't?)

I wonder whether this means that presently, mapgen will differ cross-platform (ARM vs. non-ARM) because of negative seeds being zero'd on ARM, and whether we want to preserve that if so. Oh the joys of UB...

Reply:
Casting to `s64` sounds good to me, so I applied that.